Sources/FirebaseFirestore/DocumentReference+Swift.swift

```diff
@@ -47,15 +47,23 @@ extension DocumentReference {
     return snapshot.pointee
   }
 
-  public func addSnapshotListener(_ listener: @escaping SnapshotListenerCallback) -> ListenerRegistration? {
+  public func addSnapshotListener(_ listener: @escaping SnapshotListenerCallback) -> ListenerRegistration {
     addSnapshotListener(includeMetadataChanges: false, listener: listener)
   }
 
-  public func addSnapshotListener(includeMetadataChanges: Bool, listener: @escaping SnapshotListenerCallback) -> ListenerRegistration? {
+  public func addSnapshotListener(includeMetadataChanges: Bool, listener: @escaping SnapshotListenerCallback) -> ListenerRegistration {
     let boxed = Unmanaged.passRetained(listener as AnyObject)
     let instance = swift_firebase.swift_cxx_shims.firebase.firestore.document_add_snapshot_listener(self, { snapshot, errorCode, errorMessage, pvListener in
         if let pvListener = pvListener, let callback = Unmanaged<AnyObject>.fromOpaque(pvListener).takeUnretainedValue() as? SnapshotListenerCallback {
-          callback(snapshot.pointee, errorCode.pointee, String(cString: errorMessage!))
+          func error(from errorCode: firebase.firestore.Error?) -> NSError? {
+            guard let error = errorCode, error.rawValue != 0 else { return  nil }
+            return NSError(domain: "firestore.firebase", code: Int(error.rawValue))
```

Comment:
Perhaps even better would be to make this an internal initialiser for `NSError`?

```swift
extension NSError {
  internal static func firestore(_ error: firebase.firestore.Error?) -> Error {
    ...
  }
}
```

This would be similar to `CocoaError.error(32, userInfo: ...)`

Reply:
Ah I like this idea!